torch/distributed/_functional_collectives.py

```diff
@@ -601,29 +610,49 @@ def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
         return AsyncCollectiveTensor(elem)
 
     def __repr__(self) -> str:  # type: ignore[override]
-        return f"AsyncCollectiveTensor({self.trigger_wait()})"
+        return f"AsyncCollectiveTensor({self.wait()})"
 
-    def trigger_wait(self):
+    def wait(self):
         if not self.completed:
             out = wait_tensor(self.elem)
             self.completed = True
-            return out
         else:
-            return self.elem
+            out = self.elem
 
-    def wait(self) -> torch.Tensor:
-        return wait_tensor(self.elem)
+        return self._run_callbacks(out)
+
+    def _run_callbacks(self, x):
+        for func in self.callbacks:
+            x = func(x)
+
+        self.callbacks = []
+        return x
 
     def _get_acs_underlying_tensor(self):
         """This method enables  _functional_collectives_impl to test if a tensor is an ACS"""
         return self.elem
 
     @classmethod
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        args_tuple = []
+        for arg in args:
+            if isinstance(arg, AsyncCollectiveTensor) and len(arg.callbacks) > 0:
```

Comment:
what if `len(arg.callbacks) == 0`, then we don't wait for ACT?

Reply:
hey @kwen2501 I am not familiar with what you mean by ACT, can you explain?

Reply:
yeah, right now I think we dont need to but there is some incorrectness happening I think
Ill look into it